core/src/main/java/io/grpc/internal/DelayedStream.java

```diff
@@ -56,263 +54,225 @@
-            listenerToBeCalled = listener;
-          }
-        }  // else: error already set, do nothing.
+    checkNotNull(reason, "reason");
+    boolean cancelNow;
+    ClientStreamListener listenerToClose = null;
+    synchronized (this) {
+      cancelNow = passThrough;
+      if (!cancelNow) {
+        error = reason;
+        listenerToClose = listener;
+        // TODO(ejona): Close InputStreams?
+        pendingCalls.clear();
       }
     }
-    if (listenerToBeCalled != null) {
-      Preconditions.checkState(streamToBeCancelled == null, "unexpected streamToBeCancelled");
-      listenerToBeCalled.closed(reason, new Metadata());
+    if (cancelNow) {
+      realStream.cancel(reason);
     }
-    if (streamToBeCancelled != null) {
-      streamToBeCancelled.cancel(reason);
+    if (listenerToClose != null) {
+      listenerToClose.closed(reason, new Metadata());
     }
   }
 
   @Override
   public void halfClose() {
-    if (startedRealStream == null) {
-      synchronized (this) {
-        if (startedRealStream == null) {
-          pendingHalfClose = true;
-          return;
-        }
+    delayOrExecute(new Runnable() {
+      @Override
+      public void run() {
+        realStream.halfClose();
       }
-    }
-    startedRealStream.halfClose();
+    });
   }
 
   @Override
-  public void request(int numMessages) {
-    if (startedRealStream == null) {
-      synchronized (this) {
-        if (startedRealStream == null) {
-          pendingFlowControlRequests += numMessages;
-          return;
+  public void request(final int numMessages) {
+    if (passThrough) {
+      realStream.request(numMessages);
+    } else {
+      delayOrExecute(new Runnable() {
+        @Override
+        public void run() {
+          realStream.request(numMessages);
         }
-      }
+      });
     }
-    startedRealStream.request(numMessages);
   }
 
   @Override
-  public void setCompressor(Compressor compressor) {
-    if (startedRealStream == null) {
-      synchronized (this) {
-        if (startedRealStream == null) {
-          this.compressor = compressor;
-          return;
-        }
+  public void setCompressor(final Compressor compressor) {
+    checkNotNull(compressor, "compressor");
+    delayOrExecute(new Runnable() {
+      @Override
+      public void run() {
+        realStream.setCompressor(compressor);
       }
-    }
-    startedRealStream.setCompressor(compressor);
+    });
   }
 
   @Override
-  public void setDecompressor(Decompressor decompressor) {
-    if (startedRealStream == null) {
-      synchronized (this) {
-        if (startedRealStream == null) {
-          this.decompressor = decompressor;
-          return;
-        }
-      }
-    }
-    startedRealStream.setDecompressor(decompressor);
+  public void setDecompressor(final Decompressor decompressor) {
+    checkNotNull(decompressor, "decompressor");
+    checkState(realStream != null, "How did we receive a reply before the request is sent?");
```

Comment:
... Good point. I can't find much to define what happens due to I/O. I've added an empty synchronized block to trigger happens-before.